src/Servers/Kestrel/Transport.Sockets/src/Internal/IOQueue.cs

```diff
@@ -22,35 +24,58 @@ public override void Schedule(Action<object> action, object state)
                 State = state
             };
 
+            // Enqueue prior to checking _doingWork.
             _workItems.Enqueue(work);
 
-            lock (_workSync)
+            // Fast check if already doing work
+            if (_doingWork == 0)
```

Comment:
How do we know this is faster than always attempting the interlocked exchange? Now that _doingWork is volatile, it's not completely obvious to me that this is the case.

@sebastienros Can you use your new tool to quickly run an I/O-heavy socket benchmark like PlatformPlaintext or Plaintext?

Reply:
The volatile read is just a memory read
```asm
cmp dword [esi+0x8], 0x0
jnz L0034
```
Interlocked.Exchange is more complicated and needs to coordinate/invalidate other cpu caches as its a write
```asm
mov eax, 0x1
xchg [esi+0x8], eax   ; Exchange
test eax, eax
jnz L0034
```

Reply:
> The volatile read is just a memory read which can come from cpu cache

On x86/64. These volatiles will be more expensive on ARM.